### ISSUE TITLE
String interpolation for Idris2

### ISSUE THREAD

**Issue by andrevidela** [idris-lang/Idris2#555]
TLDR: String interpolation, If you want to comment, read the 3rd section first.

[String interpolation](https://en.wikipedia.org/wiki/String_interpolation) (Also known as variable interpolation, variable expansion, or variable substitution) is a common feature of modern programming languages that allows the programmer to write a string literal and insert executable code inline for it to be automatically inserted within the string by the compiler.

Since this proposal is mostly syntactic, in order to avoid the [_Wadler effect_ ](https://wiki.haskell.org/Wadler's_Law), we are going to split it in 3 parts:

- The proposal itself
- The technical details
- The research conducted to reach the proposal

To avoid endless bike-shedding I invite you to comment on the first two, the last one aims to answer all other questions or remarks that might slow down or pollute the discussion thread with less pressing concerns.

# 1. The proposal

We propose introducing this new syntax for string literals

```scala
"this is a normal string"
s"This is an {interpolated} string"
"normal strings are {raw} strings"
s"interpolated string \{need\} to escape curly braces if you want to show them."
```

Expressions within brackets can be arbitrary complex as long as they return a type that implements `Show` :

```scala
s"This is a string with a complex value: {replace (plusCommutative n m) $ map isEven record.field}"
```

# 2. Technical details

String interpolation has 2 goals:
- Improve readability
- Inform the compiler of the user’s intention about string manipulation

While the first one can only be subjectively evaluated, the second one has a more concrete consequence for the compiler. Take this example

```haskell
let name = "Susan"
    greeting = s"hello {name}" in 
    putStrLn greeting
```

In it, the compiler can _see_ that the variable `name` comes from a string literal. From which the compiler can perform the substitution _at compile time_ rather than at runtime, outputting the code (given some inlining)

```haskell
putStrLn "hello Susan"
```

instead of 

```haskell
putStrLn ("hello " <+> "Susan")
```

In other words, the compiler can avoid runtime concatenation operation and perform those text replacement at compile-time.

## A taste of linearity

Interestingly enough, since this substitution is performed at _compile time_ it is semantically equivalent to a function call with linearity `0`, that is, the following should compile:

```haskell
let 0 name = "Susan"
    1 greeting = s"hello {name}" in
    putStrLn greeting
```

This observation leads us to believe that String Interpolation could make use of linearity (And linearity inference) to decide how to generate the correct output code. In addition, this allows the programmer to predict when the compiler will be able to perform such optimisation, and when it cannot. In the next example we can be assured the string concatenation will happen because there is no way to annotate `name` with linearity `0`

```haskell
main : IO ()
main = do name <- readLine
          putStrLn s"hello {name}"
```

In conclusion, _erased_ terms are concatenated at compile time and _linear_ and _unrestricted_ terms are compiled into classical runtime concatenation.

## `%stringlit`

Idris2 has a feature allowing string literals to be overloaded with additional meaning provided by the `FromString` interface.

```haskell
interface FromString ty where
  fromString : String -> ty
```

This feature is particularly useful when designing an embedded DSL. For example

```haskell
data URLPath : Type where
    
    Component : String -> URLPath
    (/) : URLPath -> URLPath -> URLPath
    END : URLPath

path : URLPath
path = "hello" / "world" / END
```

The `path` is equivalent to `fromString "hello" / fromString "world" / END`, or `Component "hello" / Component "world" / END` both of which are much noisier. 

The rule for `fromString` is simply that every string literal is replaced by a call to `fromString` using the literal as argument. However this rule breaks down if we try to apply it to interpolated strings. Take this example:

```haskell
let v : MyString = s"hello {"world"}" in ?rest
```

Should  it be converted to 

```haskell
let v : MyString = fromString "hello " <+> fromString "world" in ?rest
```

(In this example we assume that string interpolation is extended to work with types which have a `FromString` instance and a `Monoid` instance.)

or

```haskell
let v : MyString = fromString "hello world" in ?rest
```

Additionally, the linearity intuition breaks down too because though both  translations are valid, the first one assume a non-erased linearity and the second one makes uses of the erased linearity and there is no way to predict which one it will be.

For those reasons, making `stringlit` compatible with string interpolation is a non-goal.

# 3. Research conducted

The goal of this section is to keep the discussion focused on the topic and answer any distracting question or remark before they make it to the comment section. For this reason this section will look like a FAQ.

## Why can’t you just use string concatenation?

We can, and we’ve been doing so for years, but String manipulation is a cornerstone of software engineering. Humans communicate with text, display text, encode data in text and even replace type systems with text (see [Stringly typed](https://blog.codinghorror.com/new-programming-jargon/) programs). The ubiquity of strings manifests itself in all sorts of ways. Take the following programs that calls a shell script

```haskell
system $ idris ++ " " ++ file ++ " -o " ++ name
```

or this `show` implementation

```haskell
    show (FunDecl fc n args b) = "\n\n" ++ "(FunDecl (" ++ show fc ++ ") " ++ show n ++ " " ++ show args ++ " " ++ show b ++ ")"
```

They _work_ but since Strings are so ubiquitous we think they should be:

- Easy to read
- Easy to write
- Performant to use/works well with linearity

String concatenation fails in those three aspects. As an example, the proposed syntax changes those two examples into.

```haskell
system s"idris {file} -o {name}"
```

```haskell
    show (FunDecl fc n args b) = s"\n\n(FunDecl ({fc}) {n} {args} {b})"
```

## We already have dependent types, why not use printf like we should?

[_Type driven development in Idris_](https://www.manning.com/books/type-driven-development-with-idris) has a wonderful example of implementing a _type-safe_ printf function leveraging dependent types. It has the following type:

```haskell
printf : (fmt : String) -> PrintfType (toFormat (unpack fmt))
```

and is used like this:

```haskell
printf "hello %s" "Susan"
```

with `printf "%s %s %s"` having the type `String -> String -> String -> String`.

While this is a great exercise in dependently typed golfing, this also fails our 3 metrics, though not as badly as concatenation:

- It’s hard to read because we have to count and tell the difference between the format string and the argument values. The human brain cannot transparently replace occurrences of `%s` with the correct corresponding argument (how is it supposed to know which element it is in the argument list? One would need to count to figure it out)
- It’s not incredibly hard to write, but knowing which argument corresponds to which position isn’t as easy, specially when you get to more than 3 arguments.
- It still performs runtime concatenation and cannot take erased arguments.

## We already have Data.String.Interpolation in contrib

Again, lets reuse both our examples and our 3 criteria:

```haskell
system $ F[idris, " ", file, " -o ", name]
```

```haskell
    show (FunDecl fc n args b) = F["\n\n", "(FunDecl (", show fc, ") ", show n, " ", show args, " ", show b, ")"]
```

As we can see, the examples are shorter and more readable, this is definitely an improvement over string concatenation, however:
- typing “F” “[“ “]” instead of “”…””  isn’t actually easier
- this only replaces `++` by `,` which is welcome, but not a sufficient improvement compared to string interpolation
- This still performs runtime concatenation and doesn’t work with erased arguments
- This function is limited to `String` instead of `Show a => a`
While some of those aspects could be improved (notably the last one), the best this solution could ever achieve is:

```haskell
    show (FunDecl fc n args b) = F["\n\n", "(FunDecl (", fc, ") ", n, " ", args, " ", b, ")"]
```

which is still an unacceptable amount of noise required in order to separate terms by single spaces.

## I don’t like your choice, what about \*insert other language’s\* string interpolation syntax

Let’s explore a few of them.

### Scala

```scala
val name = "Susan"
println(s"Hello, ${name}")
```

- `s` prefix
- Dollar sign + curly braces
- Curly brace optional when used with single identifier: `$name`

### C# 

```cs
var name = "Susan"
Console.WriteLine($"Hello {name}");
```

- `$` prefix
- Curly braces

### Swift

```swift
let name = "Susan"
print("hello \(name)")
```

- No prefix
- Escaped left parenthesis

### Javascript

```javascript
var name = "Susan"
console.log(`hello ${name}`)
```

- Backticks
- Dollar sign + curly braces

### Python

```python
name = "Susan"
print(f"Hello {name}")
```

- `f` prefix
- Curly braces

### Conclusion

We needed a prefix in order to avoid clashing semantics between _raw strings_ and _interpolated strings_. Given the existing examples our choices are either `s`, `f` or `$`. The `f` prefix is unsuitable since it refers to “format” but our string interpolation proposal does not include any feature wrt to formatting (Scala has a special prefix `f` to distinguish between formatted strings and interpolated ones).  The `$` would be fine, however, `$` already has meaning in Idris as the function application operator. If we were to use it, those two functions would do different things despite looking very similar

```cs
trim $ "hello {name}  "
trim $"hello {name}   "
```

Which leaves us with the `s` prefix, it would stand for “string” which isn’t too outlandish.

As for bracketing the expressions, almost all languages use curly braces. Some of them prefix the curly braces by a `$` we would rather avoid it for the same reason we do not use it as prefix and because it makes the expression harder to type (an additional shift+4 to type on US keyboards). We feel like curly braces are the right balance of unlikeliness to appear in strings displayed to the user and ease to type. Indeed, parenthesis are ubiquitous, and square brackets are often used for list bounds. Curly braces find themselves for JSON values and Object-oriented records, but those rarely appear in typical idris code, and when they do, the string is machine generated rather than typed by a human.

It is worth noting that the curly braces are already used in Idris for records and implicit function parameters. However, those three uses are unlikely to clash in an unreadable manner, even the worst case scenario:

```haskell
nightmare : {val : Rec} -> {0 v : record { field = s"hello {record { foo = bar } rcd}" } val } -> Type
```

Isn’t hard to read because of the interpolated string but because of the record syntax.

Thanks to @shinkage who co-authored this proposal, we’re ready to answer your questions.

**Comment by gallais** [comment]
One thing I have not seen answered in this proposal:

The proper implementation of nested comments ignores closing tokens for
comments contained in a string literal. This allows you to make sure that
commenting out any well-formed program yields a well-formed program.

E.g. this is a valid single multi-line comment:

```idris2
{-

myString : String
myString = "It's a trap! -}"

-}
```

At the moment we can parse nested comments by only looking for `-}` and
`"` and switching between parsing nested comments & parsing string literals.

With your proposal allowing users to embed arbitrary code inside these
`{...}` we may need to parse arbitrary Idris expressions just to detect
when an interpolated string has ended (cf. your record-in-splice example
where a closing brace is not the end of the splice).

What about only allowing identifiers in these splices? Would forcing users
to use `let` be such an annoyance?

**Comment by ShinKage** [comment]
I think arbitrary expression should be allowed. As for regular string literals inside a block comment, we generate a single token for the comment with a simple string, thus when we find a `"` we just go to another state in the abstract lexer automata that ignores `-}` until we find another `"` and any other character is just added to the comment string without further analysis. The `stringLit` combinator used for string literal parsing, does just that, ignores everything that is not a `"`. There is no reason we could not give the same treatment for interpolated literals.

The only additional problem, with interpolated strings and not regular string literals, is nested strings (as in `f"foo {"bar"}"`). However, nested enclosed content is already dealt by nested comments, so we could adopt the same solution for nested strings.

If these changes to the lexer results in performance hits the issue pertains the lexing phase, that even if our implementation is not optimized, shoud still be asymptotically linear.

**Comment by gallais** [comment]
> However, nested enclosed content is already dealt by nested comments,

Note that for nested comments the opening & closing tokens are distinct.

**Comment by melted** [comment]
Nicely worked out proposal!

**Comment by andylokandy** [comment]
@andrevidela Did you figure out how to lex the `MiddleInterpString` yet? I'm gonna give modal lexer a try to see if it helps.

**Comment by andrevidela** [comment]
I was going to work on that today actually.

'the plan is to look at { when parsing strings and then count the number of opening braces and closing braces (in the same way we count opening comment tokens), when all the braces are closed (and the count drops to 0) it means we are done with the interpolated part and we art starting to parse `InterpolatedMiddle` until we reach another `{`.

You can give it a go since I won't be able to take a look until this evening

**Comment by ohad** [comment]
Probably needs a little bit more care, since the spliced code might contain nested comments, and those might contain `}`. There was some previous discussion about it earlier in this page ([here](https://github.com/idris-lang/Idris2/issues/555#issuecomment-673368360)).

**Comment by andylokandy** [comment]
I've written a draft for the modal lexer ([gist](https://gist.github.com/andylokandy/2b94831927dd1fba8a2ac46429321d00)). I've made it compile but not tested yet. However, it's too late here so I've to go to bed now. Hope it helps your hacking tonight @andrevidela.

**Comment by danielkroeni** [comment]
Hi, thanks for this nice feature. I found the following surprising behaviour:

```
> idris2       
     ____    __     _         ___                                          
    /  _/___/ /____(_)____   |__ \                                         
    / // __  / ___/ / ___/   __/ /     Version 0.4.0-5f3480120
  _/ // /_/ / /  / (__  )   / __/      https://www.idris-lang.org          
 /___/\__,_/_/  /_/____/   /____/      Type :? for help                    

Welcome to Idris 2.  Enjoy yourself!
Main> "\{ if True then "a" else "b" }"
Couldn't parse any alternatives:
1: Expected 'case', 'if', 'do', application or operator expression.
(Interactive):1:5--1:7
 1 | "\{ if True then "a" else "b" }"
         ^^
... (43 others)

Main> "\{ (if True then "a" else "b") }"
"a"
```

The error says that an 'if' would be expected, but at the same time it reports the 'if' as an error.

**Comment by bi-functor** [comment]
Coming to Idris from Haskell, I’m very excited for this proposal! The current state of affairs requires that the argument to an interpolation be of type `String`, forcing users to state a conversion function in each place they’d like to interpolate some other Idris value. This brings back some of the encumbrance this was intended to reduce.

One of my personal favourite interpolation libraries in Haskell is called `shakespeare`, and the way that it handles Haskell value interpolation in text is through quasiquotes and the `#{…}` syntax. However, the way it works is slightly more sophisticated, in that the interpolation accepts any value of type `(ToText a) => a`; that is, as long as it has an instance of the `ToText` class. The interpolation calls its only method, `toText :: a -> Text` on the passed value.

I think a similar mechanism could work using Idris’ interfaces. However, I strongly recommend the creation of a new interface, for example `StringInterpolate`, rather than using `Show`. In Haskell’s case, `Show` is a derivable class, which in many cases will simply print out the representation of the value as it would be written in code, for use in debugging. However, many library maintainers choose to write their own instances of `Show`, which may or may not have custom formatting or pretty-printed parts. This is especially annoying when displaying data with multiple types whose `Show` instances behave differently. To avoid the spread of this duplicitous tendency, I suggest using an interface specifically geared toward interpolation so there can be no confusion as to what it “should” or “should not” be used for.

**Comment by IFcoltransG** [comment]
I'd like to see a new interface (not `Show`) used for interpolation. My only suggestion is that the solution should be future proof, for if we later decide to add formatting parameters. For example, Rust passes a formatter struct to the `fmt` function, which it uses for converting to a string. The struct holds data from the format string, like precision, radix, whether to pad with spaces, and can be safely ignored if the implementor doesn't need those arguments. Python defers to a third method for converting to strings, `.__format__()`, which has a default implementation.

**Comment by andylokandy** [comment]
@bi-functor @IFcoltransG I contributed the interpolation string for Idris and I'm also a daily developer in Rust, I'd like to share my point of view on the `Show` interface on interpolation strings. First of all, `Show` is more like the `Debug` in Rust which **should** be derivable and should print the structure of the value. And besides that, `Pretty` is like the `Display` and has provided more powerful tools for formatting. Back to the interpolation string, `Pretty` is out of the game since we only desire a `String` rather than a `Doc`, and also `Show` is not a good choice for it because `Show` will always add quotes on Strings. So as a result, I chose to get rid of any auto conversion/formatting by interfaces but left it for programmers to convert the value into strings. There are some upsides to this approach: 1. you can specify any conversion implementation (eg. `Show`, custom function, named `Show` implementation) 2. you (or std) can define functions like `hex`, `decimal`, `bin` or more for formatting value into a string so that we don't have to implement special syntax in format strings like Rust.
